chore: Update Definition of Done (#2416)
Amsterdam/design-system: documentation/definition-of-done.md

Reviewer: General comment: there are a great many rules. That's quite a mental load for developers and maintainers to keep them in mind while they are developing or reviewing code. Many rules are covered by linters, validators and other tooling in the CI pipeline, but currently it is not clear which rules are covered and which ones are not. It would be helpful if there were lists with rules that have to be checked manually and rules for which someone can rely on automatic tooling.

# Definition of done

## Quality requirements

### For components

#### Required accessibility checks

> - Ensure focus outline is visible for interactive components.

Reviewer 2: The phrasing style is inconsistent. Some items start with the action ("Ensure focus outline is visible") while others lead with the object ("Visual changes in Chromatic are reviewed"). The passive constructions feel a little looser than the active ones.

Additionally, including the verb this sounds like we’re addressing the reviewer or the programmer. I think the points become shorter, stronger, and align more with a term like ‘requirements’ if they state them as facts or goals, e.g. ‘The focus outline is visible for interactive components’. Can we adopt that style for this doc?

> - Test on both mobile and desktop devices.

Author: The appearance adapting to the width of the screen isn't an a11y concern in itself, more of a design one. 

How about: 

- Test on smaller screens: verify content does not overflow or get cut off.
- Test pointer (mouse), keyboard, and touch interactions. Verify custom gestures do not conflict with native ones, for example a horizontal swipe to scroll conflicting with the browser's swipe-to-navigate gesture.

- Verify visibility and usability when zoomed up to 200%.
- Verify visibility and usability when adjusting the text size to the maximum.
- Confirm that color is not the sole means of conveying information.
- Validate the component using [Lighthouse](https://developer.chrome.com/docs/lighthouse/overview), [WAVE](https://wave.webaim.org/standalone), and [the Markup Validation Service](https://validator.w3.org/).

> - Test with a screen reader.

Reviewer 2: This is also quite open.

- If the component includes animation, verify it respects the `prefers-reduced-motion` setting.
- Apply the provided CSS snippet and confirm that all elements are still rendered correctly, adhering to WCAG 1.4.12. In Chrome, you can utilize the Stylus plugin for easy implementation.

```
* {
  line-height: 1.5 !important;
  letter-spacing: 0.12em !important;
  word-spacing: 0.16em !important;
}

p {
  margin-bottom: 2em !important;
}
```

#### Additional accessibility checks

- Translate the component into a right-to-left (RTL) language and verify that all content, including `aria-label` if used, is translated and displays correctly in RTL.
- Check the print preview for correct rendering.
- Check if the component still makes sense when CSS fails to load.
- Ensure the component is still functional without JavaScript enabled.

> - Test in forced colors mode.

Reviewer 2: Can we explain a bit more what to look out for?

#### Automated testing

- Unit tests are written and pass.
- A test story renders all component variants for visual regression testing.
- Visual changes in Chromatic are reviewed and approved.
- ESLint, Stylelint, and Prettier report no errors.

### For pages

> - All required accessibility checks should pass.

Reviewer 2: ```suggestion
- All required accessibility checks pass.
```

  - All meaningful images and icons have descriptive alternative text.
  - Decorative images and icons are hidden from assistive technology.
- `prefers-reduced-motion` is used if it includes animation.
- Content does not overflow or get cut off when the provided CSS snippet is applied. See [WCAG 1.4.12](https://www.w3.org/TR/WCAG22/#text-spacing). In Chrome, the [Stylus plugin](https://chromewebstore.google.com/detail/stylus/clngdbkpkpeebahjckkjfobafhncgmne) makes this easier.

```
* {
  line-height: 1.5 !important;
  letter-spacing: 0.12em !important;
  word-spacing: 0.16em !important;
}

p {
  margin-bottom: 2em !important;
}
```

#### Additional accessibility checks

- All content, including `aria-label` if used, displays correctly in a right-to-left (RTL) language when translated.
- The print preview renders the same as on a smaller screen, with the exception of print-specific styles.
- The component is usable when CSS fails to load: unnecessary elements are not rendered, and visual elements like images are not shown overly large.
- The component degrades gracefully when JavaScript is unavailable or fails to load:
  - Static content is visible and readable.

>   - Any functionality that can be implemented in HTML/CSS alone still works (e.g., a <details> element for disclosure, native <select> for dropdowns).

Reviewer: The HTML in this Markdown document breaks the layout and the rendering of further content.

See: https://github.com/Amsterdam/design-system/blob/1464166fbbb0e1b13aef495067ff8975d983c201/documentation/definition-of-done.md

>   - Any functionality that can be implemented in HTML/CSS alone still works (e.g., a <details> element for disclosure, native <select> for dropdowns).

Reviewer: ```suggestion
  - Any functionality that can be implemented in HTML/CSS alone still works (e.g., a `<details>` element for disclosure, native `<select>` for dropdowns).
```

  - The component does not render a broken or empty state (e.g., a spinner that never resolves, or a blank container where content should be).
- Works correctly in forced colors mode:
  - All text and interactive elements remain visible and distinguishable.
  - Focus indicators are visible.
  - Icons and graphical elements that convey meaning are visible and distinguishable.
  - Borders and outlines are present where needed to distinguish UI regions.

#### Automated testing

> - Unit tests are written and pass.

Reviewer 2: ```suggestion
- Unit tests have been written and pass.
```

- A test story renders all component variants for visual regression testing.
- Visual changes in Chromatic are reviewed and approved.

> - ESLint, Stylelint, and Prettier report no errors.

Reviewer: Personally, I would prefer if we'd move away from Prettier to [ESLint Stylistic](https://eslint.style/).

Why?

Well, read [their explanation](https://eslint.style/guide/why#linters-vs-formatters).

And for me added benefits are:

- Less tooling in the project and developer's IDEs that does sort of the same thing;
- It's simpler: you're dealing with just one config file and one API.

  - All CSS declarations are necessary and have an effect.
  - All CSS classes defined are actually used in the HTML.
  - All classes used in the HTML are defined in the CSS.
- **Token usage:**
  - All tokens used are defined.
  - No unused tokens are defined.
- **Unit tests:**
  - Ensure tests use the correct ARIA roles.
  - For enum props, all variants are covered in unit tests.
- **Exports:**
  - Add or update exports in `index.*` files.
  - Export the main component and its type. Also export the types of any subcomponents.
- **Stories:**
  - Add or update Storybook stories for new or changed components.

## Code conventions

- You’ve followed [the contribution guidelines](../CONTRIBUTING.md).
- You’ve followed the [NL Design System developer guidelines](https://nldesignsystem.nl/handboek/developer/introductie/).
- CSS follows [the CSS coding conventions](../packages/css/documentation/coding-conventions.md).
- React code follows [the React coding conventions](../packages/react/documentation/coding-conventions.md).

## Documentation

- Props are typed with JSDoc descriptions.
- A component README is written (documentation in English, example content in Dutch).

> - Storybook stories exists for all component variants that are relevant to our users.

Reviewer 2: ```suggestion
- Storybook stories exist for all component variants relevant to our users.
```
